[FLINK-13245][network] Fix the bug of resource leak while releasing partition and view reader

zhijiangW opened the pull request:
## What is the purpose of the change

*The view reader release and the relevant partition release based on reference counter in `ReleaseOnConsumptionResultPartition` do not work correctly within expection, which would result in resource leak. In order to fix this issue we need to adjust the following parts:*

*1. On producer side the netty stack handles the `CancelPartitionRequest` via finding the proper view from available queue atm. But the corresponding view might not always be available, so it is reasonable to trace it from all readers queue instead.*

*2. We could always call `NetworkSequenceViewReader#notifySubpartitionConsumed` along with `NetworkSequenceViewReader#releaseAllResources` in any cases. Then the reference counter based release could work correctly in `ReleaseOnConsumptionResultPartition` and the logic is simple in netty stack.*

*3. In the implementation of `ReleaseOnConsumptionResultPartition#onConsumedSubpartition`, if one subpartition is consumed multiple times, the reference counter is also decreased multiple times. Then it would cause the whole result partition released after only partial subpartitions consumed multiple times. A boolean tag is added for indicating the state of every subpartition.*

## Brief change log

  - *Fix the logic in handling `CancelPartitionRequest` in `PartitionRequestQueue`*
  - *Always calling `notifySubpartitionConsumed` while calling `releaseAllResources`*
  - *Remove redundant bookkeeping for already canceled input channel IDs*
  - *Add the boolean tag for every subpartition to avoid decrease reference counter multiple times for one subpartition in `ReleaseOnConsumptionResultPartition`.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests for verifying the partition and view are both released while receiving `CancelPartitionRequest` for the case of available/unavailable view.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhijiangW commented:
@flinkbot attention @zentol @azagrebin

zhijiangW commented:
@zentol , for the case of `SlotCountExceedingParallelismTest` it would not generate `ReleaseOnConsumptionResultPartition` because of blocking partition. Then the reference counter is not actually used for releasing partitions, but the partitions would finally be release via `RegionPartitionReleaseStrategy`.

This PR change seems only make reference counter work for pipelined partitions. I run the `SlotCountExceedingParallelismTest` locally in Mac system, and it does not have file leaks before/after this fix. If in windows system it has file leaks then I guess it might be relevant with mmap mechanism in different systems.

In addition, how do you test as mentioned above? Do you mean shutting down the cluster/job during test running? If the tests are not running finished, then it would indeed leftover `.channel` files.

NicoK commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueueTest.java:
```diff
@@ -348,6 +376,68 @@ public void testEnqueueReaderByNotifyingBufferAndCredit() throws Exception {
+		assertThat(partition.isReleased(), is(true));
+		for (ResultSubpartition subpartition : partition.getAllPartitions()) {
+			assertThat(subpartition.isReleased(), is(true));
+		}
```
Similar to above, if you don't want messages like this:
```
java.lang.AssertionError: 
Expected: is <true>
     but: was <false>
```
then rather do these checks another way:
```
		assertThat(queue.getAvailableReaders(), not(contains(reader)));
		// the partition and its reader view should all be released
		assertTrue("reader not released", reader.isReleased());
		assertTrue("partition not released", partition.isReleased());
		for (ResultSubpartition subpartition : partition.getAllPartitions()) {
			assertTrue("subpartition not released", subpartition.isReleased());
		}
```
(FYI: I didn't use `assertThat(reader, hasProperty("released", is(true)));` because it would be prone to be forgotten during refactorings)

zentol commented:
@zhijiangW My bad, your fix did solve the problem of files leaking after a job terminates. The breakpoint I set likely paused the cluster before it could execute the cleanup operations.

NicoK commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueueTest.java:
```diff
@@ -348,6 +376,68 @@ public void testEnqueueReaderByNotifyingBufferAndCredit() throws Exception {
+		channel.close();
+	}
+
+	private static ResultPartition createPartitionWithFilledData(ResultPartitionManager partitionManager) throws Exception {
```
It is, however, a very special helper method and (in the current form) only useful for this one test. However, it keeps the test clean with respect to the logic that it has. Nothing to change now(!), but if used with more variants, this could move into `PartitionTestUtils` as well.

zhijiangW replied:
Agree, it could be moved to `PartitionTestUtils` if to be used in other relevant classes future.
The current motivation is making the existing test logic clean and it might also be used for other new unit tests in this class future.

StephanEwen commented:
I am wondering why we distinguish cases where we release with/without consumption notification.
To me, all release should include a consumption notification - similar to what @azagrebin raised in the JIRA discussion.

The reason is that `notifySubpartitionConsumed()` has a semantic of "a consumption attempt finished".
And all release cases also indicate that.

I understand that this is not exactly the behavior we had before, but I would argue that the behavior before was actually incorrect and we should fix this now.

StephanEwen commented:
@zhijiangW @NicoK @azagrebin @zentol 

I addressed the review comments, put some minor changes on top to make consumption notifications idempotent. That allows also for some more slight simplification/cleanup. Will open a PR with this as soon as I the tests passed once on Travis.

https://github.com/StephanEwen/flink/commits/network

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -181,19 +179,18 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object msg) throws Exc
 
-			allReaders.remove(toCancel);
+			// remove reader from queue of all readers and release its resource
+			NetworkSequenceViewReader toRelease = allReaders.remove(toCancel);
```
The `toRelease` could be null if it was already released on server side (any network exceptions) before receiving cancel message from client side.

I think it might be proper to ignore non-existing readers here. And it also makes sense to remove `PartitionRequestQueue#released`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PartitionTestUtils.java:
```diff
@@ -158,4 +159,11 @@ public ResultPartitionID getResultPartitionID() {
+			partition.addBufferConsumer(createFilledBufferConsumer(bufferSize, bufferSize), 0);
+		}
+		partition.finish();
+	}
```
Actually it could be done and I indeed ever considered these parts while implementation.
It might be a separate hotfix commit for refactoring these previous tests and I did not want to touch them at that moment in limited time to block release.

zhijiangW replied:
In addition, these three tests are focusing on different classes actually, which are `ResultPartition`, `ResultSubpartition` and `BoundedData`, so they are not totally duplications. If we want to unify them, we have to also refactor the relevant unit tests. I do not think it is necessary to do so now.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueueTest.java:
```diff
@@ -348,6 +376,68 @@ public void testEnqueueReaderByNotifyingBufferAndCredit() throws Exception {
+			.build();
+
+		partitionManager.registerResultPartition(partition);
+		PartitionTestUtils.writeBuffers(partition, 1, BUFFER_SIZE);
```
I think both ways are ok here, since we already had this helper method.

zhijiangW commented:
@StephanEwen  thanks for helping address above comments and some further refactoring.

I have only one concern for the issue of `notifySubpartitionConsumed`. I agree that it would make the logic simple on server side to not distinguish during release resources, and it actually makes sense to do so now. But this distinguish might have valuable effects in future. E.g. in the case of any local network exceptions on server side, if the server releases the reader/view together with the subpartition/partition, and then notify the client side about the `ErrorResponse`. That means there might be no chances for creating views for partition any more because it has already been removed from `ResultPartitionManager`. In current way, only reader/view is released and the client still have the opportunity to decide whether to request partitions again after receiving `ErrorResponse` in future.

The key point is whether trigger calling `notifySubpartitionConsumed` via client's specific confirmation or server could do that in any ways. The previous assumption is that it should be done in client's specific confirmation.

I would pick the relevant commits from @StephanEwen 's branch excepting for whether to always call `notifySubpartitionConsumed` during release.

zhijiangW commented:
After some offline discussions with @StephanEwen  and @azagrebin . We could make the current netty stack logic simple for always calling `notifySubpartitionConsumed` while releasing partition's reader/view in any cases.

For pipelined mode, the current assumption is that one subpartition would only have one consumption attempt. 

For blocking mode, one subpartition could have multiple readers/views and the partition release is up to scheduler/JM.

In release-1.10 we could further refactor the current process as follows:

- In netty handler stack, it would only call view/reader.releaseResources in the current cases.

- In reader/view stack, it would further notify the parent of the released state, maybe still via the method of `notifySubpartitionConsumed`.

- In partition stack, it could decide whether to release itself or not based on some factors. E.g. for pipelined partition it could release partition as well. And for blocking partition it would do nothing.

zhijiangW commented:
@NicoK I already addressed your above comments and picked some additional cleanup commits from  @StephanEwen's branch. Could you double check if there are any issues now?

StephanEwen commented:
From my side, +1 to commit this

zhijiangW commented:
I adjusted the PR description based on the latest modifications, and all the commits would be squashed after approval finally and the commit content would be adjusted as well.

StephanEwen commented:
Thanks, looks good.
Merging this...

zhijiangW commented:
Thanks for merging, closed it.
